internal/slicer/slicer.go

```diff
@@ -235,38 +242,54 @@ func Run(options *RunOptions) (*Report, error) {
 		reader.Close()
 		packages[slice.Package] = nil
 		if err != nil {
-			return nil, err
+			return err
 		}
 	}
 
-	// Create new content not coming from packages.
-	done := make(map[string]bool)
+	// Create new content not coming from packages excluding GeneratePath(s).
+	// First group them by their relative path. Then create them and attribute
+	// them to the appropriate slices.
+	relPaths := map[string][]*setup.Slice{}
 	for _, slice := range options.Selection.Slices {
 		arch := archives[slice.Package].Options().Arch
 		for relPath, pathInfo := range slice.Contents {
 			if len(pathInfo.Arch) > 0 && !slices.Contains(pathInfo.Arch, arch) {
 				continue
 			}
-			if done[relPath] || pathInfo.Kind == setup.CopyPath || pathInfo.Kind == setup.GlobPath {
+			if pathInfo.Kind == setup.CopyPath || pathInfo.Kind == setup.GlobPath ||
+				pathInfo.Kind == setup.GeneratePath {
 				continue
 			}
-			done[relPath] = true
-			data := pathData{
-				until:   pathInfo.Until,
-				mutable: pathInfo.Mutable,
-			}
-			addKnownPath(knownPaths, relPath, data)
-			targetPath := filepath.Join(targetDir, relPath)
-			entry, err := createFile(targetPath, pathInfo)
-			if err != nil {
-				return nil, err
+			relPaths[relPath] = append(relPaths[relPath], slice)
+		}
+	}
+	for relPath, slices := range relPaths {
+		until := setup.UntilMutate
+		for _, slice := range slices {
+			if slice.Contents[relPath].Until == setup.UntilNone {
+				until = setup.UntilNone
+				break
 			}
+		}
+		pathInfo := slices[0].Contents[relPath]
```

Comment:
The diff above feels a bit suspect: it changed an iteration that happened just once to an iteration that happens once to append all iterated elements to a map=>list, to then almost immediately discard that list and work with the very first item, with almost everything else somewhat untouched. The only new thing seems to be the logic above regarding until, but it looks like this could easily be integrated into the original code instead of the reshaping that took place here.

Reply:
Here the important bit is the attribution to the slices. If two slices declare the same {text: } file we want to create it once but both slices should be recorded as "owners" in the manifest. Hence why we do below (note `slices`):
```go
		// Do not add paths with "until: mutate".
		if pathInfo.Until != setup.UntilMutate {
			for _, slice := range slices {
				err = report.Add(slice, entry)
				if err != nil {
					return err
				}
			}
		}
```
The only reason we use `slices[0]` is that `setup` has already validated the release so all of them have the same mutability attribute (but not `until`). Maybe we need to start coding this assumptions as asserts instead so it is clear in the code and more defensive.

Reply:
I see, you are right. Thanks.

Reply:
Discussed offline: add a comment saying that it is okay to use the data for the first slice because we have validated the release and the paths have to be compatible.

Reply:
Done in https://github.com/canonical/chisel/pull/142/commits/aaf70b1945182a09bf7b76f2ee7595302cdbd2d8.